src/tokenizer.cc

```diff
@@ -0,0 +1,707 @@
+                return 21;
+            case 'w':
+                return 22;
+            case 'x':
+                return 23;
+            case 'y':
+                return 24;
+            case 'z':
+                return 25;
+            default:
+                return 0; // Should not reach here
+        }
+    }
+    return 0; // Default return value
+}
+
+/**
+ * Retrieves the string value of the current token
+ *
+ * @param none
+ * @return Reference to the token's string value if it contains a string,
+ *         otherwise returns reference to empty string
+ *
+ * Used to access string literals and keywords from the token_data
+ * Returns empty string if token_data doesn't contain a string value
+ */
+std::string_view Tokenizer::get_string() const {
+    static const std::string empty;
+    if (std::holds_alternative<std::string>(token_data_)) {
+        return std::get<std::string>(token_data_);
+    }
+    return empty;
+}
+
+/**
+ * Retrieves the numeric value of the current token
+ *
+ * @param none
+ * @return Integer value if token contains a number, otherwise returns 0
+ *
+ * Used to access numeric literals from the token_data
+ * Returns 0 if token_data doesn't contain an integer value
+ */
+int Tokenizer::get_num() const {
+    if (std::holds_alternative<int>(token_data_)) {
+        return std::get<int>(token_data_);
+    }
+    return 0; // Default return value
+}
+
+/**
+ * Advances to the next token in the input stream
+ *
+ * @param none
+ * @return none
+ *
+ * Skips whitespace (spaces and tabs, but preserves newlines) and updates
+ * current_token_ by calling get_next_token()
+ * Does nothing if tokenizer has already reached EOF
+ */
+void Tokenizer::next_token() {
+    if (finished()) {
+        return;
+    }
+
+    // Skip spaces and tabs, but not newlines
+    while (io_->current() == ' ' || io_->current() == '\t') {
+        io_->next();
+    }
+
+    current_token_ = get_next_token();
+}
+
+/**
+ * Skips all characters until the end of the current line
+ * @param none
+ * @return none
+ */
+void Tokenizer::skip_to_eol() {
+    while (io_->current() != '\n' && io_->current() != '\r' &&
+           io_->current() != EOF && io_->current() != '\0') {
+        io_->next();
+    }
+    if (io_->current() == '\r') {
+        io_->next();
+        if (io_->current() == '\n') {
+            io_->next();
+        }
+    } else if (io_->current() == '\n') {
+        io_->next();
+    }
+    current_token_ = get_next_token();
+}
+
+/**
+ * Reads and classifies the next token from the input stream
+ *
+ * @param none
+ * @return The TokenType of the next token in the input
+ *
+ * Updates the input stream position (io_) and token_data_ as needed
+ * Returns ERROR for unrecognized characters
+ * Logs debug information about character processing if DEBUG_LOG is enabled
+ */
+
+Tokenizer::TokenType Tokenizer::get_next_token() {
+    int c = io_->current();
+    DEBUG_LOG("get_next_token processing char: '"
```

Comment:
glad you're using debug logging, also using `assert` can help a lot in development builds, and you can turn them off in production builds